libraries/standard/http/src/http_client.c

```diff
@@ -29,23 +34,329 @@ static HTTPStatus_t _sendHttpBody( const HTTPTransportInterface_t * pTransport,
+    if( ( initHeadersLen + toAddLen ) <= pRequestHeaders->bufferLen )
+    {
+        if( hasTrailingLine == 1u )
+        {
+            pRequestHeaders->headersLen = initHeadersLen - HTTP_HEADER_LINE_SEPARATOR_LEN;
+        }
+
+        /* Write "Field: Value \r\n\r\n" to headers. */
+        memcpy( pBufferCur, pField, fieldLen );
+        pBufferCur += fieldLen;
+        memcpy( pBufferCur, HTTP_HEADER_FIELD_SEPARATOR,
+                HTTP_HEADER_FIELD_SEPARATOR_LEN );
+        pBufferCur += HTTP_HEADER_FIELD_SEPARATOR_LEN;
+        memcpy( pBufferCur, pValue, valueLen );
+        pBufferCur += valueLen;
+        memcpy( pBufferCur, HTTP_HEADER_LINE_SEPARATOR, HTTP_HEADER_LINE_SEPARATOR_LEN );
+        pBufferCur += HTTP_HEADER_LINE_SEPARATOR_LEN;
+        memcpy( pBufferCur, HTTP_HEADER_LINE_SEPARATOR, HTTP_HEADER_LINE_SEPARATOR_LEN );
+        pRequestHeaders->headersLen += toAddLen;
+        returnStatus = HTTP_SUCCESS;
+    }
+    else
+    {
+        IotLogErrorWithArgs( "Unable to add header in buffer: ",
+                             "Buffer has insufficient memory: ",
+                             "RequiredBytes=%d, RemainingBufferSize=%d",
+                             toAddLen,
+                             ( pRequestHeaders->bufferLen - pRequestHeaders->headersLen ) );
+        returnStatus = HTTP_INSUFFICIENT_MEMORY;
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static HTTPStatus_t _writeRequestLine( HTTPRequestHeaders_t * pRequestHeaders,
+                                       const char * pMethod,
+                                       size_t methodLen,
+                                       const char * pPath,
+                                       size_t pathLen )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    uint8_t * pBufferCur = pRequestHeaders->pBuffer;
+    size_t toAddLen = methodLen +                 \
+                      SPACE_CHARACTER_LEN +       \
+                      pathLen +                   \
+                      SPACE_CHARACTER_LEN +       \
+                      HTTP_PROTOCOL_VERSION_LEN + \
+                      HTTP_HEADER_LINE_SEPARATOR_LEN;
+
+    if( ( toAddLen + pRequestHeaders->headersLen ) > pRequestHeaders->bufferLen )
+    {
+        returnStatus = HTTP_INSUFFICIENT_MEMORY;
+    }
+
+    if( returnStatus == HTTP_SUCCESS )
+    {
+        /* Write "<METHOD> <PATH> HTTP/1.1\r\n" to start the HTTP header. */
+        memcpy( pBufferCur, pMethod, methodLen );
+        pBufferCur += methodLen;
+        memcpy( pBufferCur, SPACE_CHARACTER, 1 );
+
+        pBufferCur += SPACE_CHARACTER_LEN;
+
+        /* Use "/" as default value if <PATH> is NULL. */
+        if( ( pPath == NULL ) || ( pathLen == 0 ) )
+        {
+            /* Revise toAddLen to contain <HTTP_EMPTY_PATH_LEN> instead. */
+            toAddLen = ( toAddLen - pathLen ) + HTTP_EMPTY_PATH_LEN;
```

Comment:
Since pathLen is 0 in this case, how do you know that the buffer is big enough for the addition of HTTP_EMPTY_PATH?